Adds concrete types distinction to JSON ABI. (#599)
FuelLabs/fuel-specs: src/abi/json-abi-format.md

  - `"concreteTypeId"`: a unique string hash based ID. Generated as specified in [Hash Based Ids](./hash-based-ids.md).
  - `"metadataTypeId"`: the the _type metadata declaration_ ID of this type, if the type metadata has components or is generic, otherwise non existent.
  - `"typeArguments"`: an array of _type concrete declarations_ hash based IDs of the type parameters of the type, if the type is generic, otherwise non existent.
- `"typesMetadata"`: an array describing all the _type metadata declarations_ used (or transitively used) in the ABI. Each _type metadata declaration_ is a JSON object that contains the following properties:
  - `"type"`: a string representation of the _type metadata declaration_. The section [JSON ABI Format for Each Possible Metadata Type Declaration](#json-abi-format-for-each-possible-metadata-type-declaration) specifies the format for each possible type.
  - `"metadataTypeId"`: a unique integer ID.
  - `"components"`: an array of the components of a given type, if any, otherwise non existent. Each component is a _type application_ represented as a JSON object that contains the following properties:
    - `"name"`: the name of the component.

>     - `"typeId"`: the _type metadata declaration_ ID or _type concrete declaration_ hash based ID of the type of the component.

Reviewer: This `typeId` is currently _always_ a string, even though it might point to a `metadataTypeId` which is an integer. If we were to make it a _number or string_, we'd know based on the value's type if it's a metadata type or a concrete type and we wouldn't have to do any string parsing. It's a minor change proposal, though, and one can live with the current spec + string parsing as well.

  - `"name"`: the name of the function
  - `"inputs"`: an array of objects that represents the inputs to the function (i.e. its parameters). Each input is a _type application_ represented as a JSON object that contains the following properties:
    - `"name"`: the name of the input.
    - `"concreteTypeId"`: the _type concrete declaration_ hash based ID of the type of the input.
  - `"output"`: the _type concrete declaration_ hash based ID of the type being returned by the function.
- `"loggedTypes"`: an array describing all instances of [`log`](../fuel-vm/instruction-set.md#log-log-event) or [`logd`](../fuel-vm/instruction-set.md#logd-log-data-event) in the contract's bytecode. Each instance is a JSON object that contains the following properties:
  - `"logId"`: a string containing the 64bit hash based decimal ID calculated from the first 8 bytes of the `sha256` of a string that represents the type logged as defined in [Hash Based Ids](./hash-based-ids.md). The [`log`](../fuel-vm/instruction-set.md#log-log-event) and [`logd`](../fuel-vm/instruction-set.md#logd-log-data-event) instructions must set their `$rB` register to that ID.

>   - `"loggedType"`: the _type concrete declaration_ hash based ID of the value being logged.

Reviewer: I think renaming this to `concreteTypeId` to follow the same pattern as function inputs makes sense.

- `"messagesTypes"`: an array describing all instances of [`smo`](../fuel-vm/instruction-set.md#smo-send-message-to-output) in the contract's bytecode. Each instance is a JSON object that contains the following properties:
  - `"message_id"`: a unique string ID.

>   - `"messageDataType"`: the _type concrete declaration_ hash based ID of the message data being sent.

Reviewer: Same as [comment on `loggedType`](https://github.com/FuelLabs/fuel-specs/pull/599/files#r1680639997).

- `"specVersion"`: a string representing the version pointing to this document versioning. `specVersion` enables the reader of the JSON ABI to find the correct specification for that file, this can be done by comparing it to value in [spec version](#spec-version).
- `"encodingVersion"`: a string representing the version of the `ABIEncode` and `ABIDecode` used in this program.
- `"programType"`: a string that can be `"script"`, `"contract"`, `"predicate"`, `"library"`. This is used by the SDK to generate types without having to manually specify the program type.
- `"concreteTypes"`: an array describing all the _type concrete declarations_ used (or transitively used) in the ABI. Each _type concrete declaration_ is a JSON object that contains the following properties:
  - `"type"`: a string representing the type, the `sha256` of this string generates the `concreteTypeId`.
  - `"concreteTypeId"`: a unique string hash based ID. Generated as specified in [Hash Based Ids](./hash-based-ids.md).

>   - `"metadataTypeId"`: the the _type metadata declaration_ ID of this type, if the type metadata has components or is generic, otherwise non existent.

Reviewer 2: ```suggestion
  - `"metadataTypeId"`: the _type metadata declaration_ ID of this type, if the type metadata has components or is generic, otherwise nonexistent.
```

  - `"typeArguments"`: an array of _type concrete declarations_ hash based IDs of the type parameters of the type, if the type is generic, otherwise non existent.

> - `"typesMetadata"`: an array describing all the _type metadata declarations_ used (or transitively used) in the ABI. Each _type metadata declaration_ is a JSON object that contains the following properties:

Reviewer: I suggest to rename this to `metadataTypes` to follow the `concreteTypes` nomenclature.

  - `"type"`: a string representation of the _type metadata declaration_. The section [JSON ABI Format for Each Possible Metadata Type Declaration](#json-abi-format-for-each-possible-metadata-type-declaration) specifies the format for each possible type.
  - `"metadataTypeId"`: a unique integer ID.
  - `"components"`: an array of the components of a given type, if any, otherwise non existent. Each component is a _type application_ represented as a JSON object that contains the following properties:
    - `"name"`: the name of the component.
    - `"typeId"`: the _type metadata declaration_ ID (number) or _type concrete declaration_ hash based ID (string) of the type of the component.
    - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the component, if the type is generic, otherwise non existent. Each _type argument_ is a _type application_ represented as a JSON object that contains the following properties:
      - `"typeId"`: the _type metadata declaration_ ID (number) or _type concrete declaration_ hash based ID (string) of the type of the component.

>       - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the _type argument_, if the type is generic, otherwise non existent. The format of the elements of this array recursively follows the rules described in this section.

Reviewer: ```suggestion
```
A metadata type can have components and those components point to types via the `typeId` property. Below is a list of the possible types a component can point to and the state of `typeArguments` based on it:
1. Concrete type - the component won't have `typeArguments` ([concrete type example](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR405)),
2. Non-generic metadata type - the component won't have `typeArguments` ([non-generic metadata type example](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR401)),
3. Generic metadata type - the component will have `typeArguments` that point to generics, and those `typeArguments` won't have any `typeArguments` themselves ([generic metadata type example](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR561))

Based on this, it seems to me that the `typeArguments` property shouldn't exist on a component's type argument anymore. What do you think?

>     - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the component, if the type is generic, otherwise non existent. Each _type argument_ is a _type application_ represented as a JSON object that contains the following properties:
>       - `"typeId"`: the _type metadata declaration_ ID (number) or _type concrete declaration_ hash based ID (string) of the type of the component.
>       - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the _type argument_, if the type is generic, otherwise non existent. The format of the elements of this array recursively follows the rules described in this section.

Reviewer: ```suggestion
    - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the component, if the type is generic, otherwise non existent. The values here must exist in the in the `typeParameters` array.
```

A metadata type can have components and those components point to types via the `typeId` property. Below is a list of the possible types a component can point to and the state of `typeArguments` based on it:
1. Concrete type - the component won't have `typeArguments` ([concrete type example](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR405)),
2. Non-generic metadata type - the component won't have `typeArguments` ([non-generic metadata type example](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR401)),
3. Generic metadata type - the component will have `typeArguments` that point to generics, and those `typeArguments` won't have any `typeArguments` themselves ([generic metadata type example](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR561))

Based on this, it seems to me that the `typeArguments` property shouldn't exist on a component's type argument anymore and, further, the `component.typeArguments` array can be simplified to be a number array. What do you think?

Reviewer: I see, thanks for the example. Below I have written an alternative which would still enable the specification to be like in the initial comment above. I wrote it for completeness sake, to get the idea out. You've answered my initial comment, so feel free to resolve the conversation.

---

I think that this can be mitigated by generating the ABI like below, where every implementation of a generic has its concrete type. That would simplify the specification by making it as suggested above, but there are implications on the ABI json size and I don't know what that entails later down the line in relation to the fuel ecosystem, if ABIs will be stored somewhere, etc.

```json
{
  "programType": "contract",
  "specVersion": "1",
  "encodingVersion": "1",
  "concreteTypes": [
    {
      "type": "()",
      "concreteTypeId": "2e38e77b22c314a449e91fafed92a43826ac6aa403ae6a8acb6cf58239fbaf5d"
    },
    {
      "type": "u64",
      "concreteTypeId": "d53e75c225dd2bb336e218026cbe306ad53e75c225dd2bb336e218026cbe306a",
    },
    {
      "type": "struct MyStruct2<b256>",
      "concreteTypeId": "ecedd677e41b0b70105bf275d1cb6645ecedd677e41b0b70105bf275d1cb6645",
      "metadataTypeId": 2,
      "typeArguments": ["d53e75c225dd2bb336e218026cbe306ad53e75c225dd2bb336e218026cbe306a"]
    },
    {
      "type": "struct MyStruct2<MyStruct2<b256>>",
      "concreteTypeId": "73458764501baff875aa17c3f53fa79273458764501baff875aa17c3f53fa792",
      "metadataTypeId": 2,
      "typeArguments": ["ecedd677e41b0b70105bf275d1cb6645ecedd677e41b0b70105bf275d1cb6645"]
    },
    {
      "type": "struct MyStruct",
      "concreteTypeId": "392d58c694d2d91f3025f2bccfadacf2a105936f5da881b0899185d49f264522",
      "metadataTypeId": 1
    }
  ],
  "typesMetadata": [
    {
      "type": "generic T",
      "metadataTypeId": 0
    },
    {
      "type": "struct MyStruct",
      "metadataTypeId": 1,
      "components": [
        {
          "name": "v",
          "typeId": "73458764501baff875aa17c3f53fa79273458764501baff875aa17c3f53fa792",
        }
      ]
    },
    {
      "type": "struct MyStruct2",
      "metadataTypeId": 2,
      "components": [
        {
          "name": "v",
          "typeId": 0
        }
      ],
      "typeParameters": [
        0
      ]
    }
  ],
  "functions": [
    {
      "inputs": [
        {
          "name": "a1",
          "concreteTypeId": "392d58c694d2d91f3025f2bccfadacf2a105936f5da881b0899185d49f264522"
        }
      ],
      "name": "test_function",
      "output": "2e38e77b22c314a449e91fafed92a43826ac6aa403ae6a8acb6cf58239fbaf5d",
      "attributes": null
    }
  ],
  "loggedTypes": [],
  "messagesTypes": [],
  "configurables": []
}
```

    - `"name"`: the name of the attribute.
    - `"arguments"`: an array of attribute arguments.
- `"loggedTypes"`: an array describing all instances of [`log`](../fuel-vm/instruction-set.md#log-log-event) or [`logd`](../fuel-vm/instruction-set.md#logd-log-data-event) in the contract's bytecode. Each instance is a JSON object that contains the following properties:
  - `"logId"`: a string containing the 64bit hash based decimal ID calculated from the first 8 bytes of the `sha256` of a string that represents the type logged as defined in [Hash Based Ids](./hash-based-ids.md). The [`log`](../fuel-vm/instruction-set.md#log-log-event) and [`logd`](../fuel-vm/instruction-set.md#logd-log-data-event) instructions must set their `$rB` register to that ID.
  - `"concreteTypeId"`: the _type concrete declaration_ hash based ID of the value being logged.
- `"messagesTypes"`: an array describing all instances of [`smo`](../fuel-vm/instruction-set.md#smo-send-message-to-output) in the contract's bytecode. Each instance is a JSON object that contains the following properties:

>   - `"message_id"`: a unique string ID.

Reviewer 3: My apologies for the post-merge feedback, but was there any particular reason why this is snake case? @esdrubal

```json
{
  "concreteTypes": [
    {
      "type": "([str[5]; 3], bool, b256)",
      "concreteTypeId": "625531542be70834dd127e771101ac1014111718451bfae996d97abe700c66a5",
      "metadataTypeId": 1,
    },
    {
      "type": "str[5]",
      "concreteTypeId": "84877f6e98274b9e4721db68b4c0bdb9e52b8e9572c5bd7811c07a41ced882c7",
    },
    {
      "type": "struct MyStruct",
      "concreteTypeId": "392d58c694d2d91f3025f2bccfadacf2a105936f5da881b0899185d49f264522",
      "metadataTypeId": 4,
    },
    {
      "type": "u64",
      "concreteTypeId": "1506e6f44c1d6291cdf46395a8e573276a4fa79e8ace3fc891e092ef32d1b0a0"
    },
    {
      "type": "b256",
      "concreteTypeId": "7c5ee1cecf5f8eacd1284feb5f0bf2bdea533a51e2f0c9aabe9236d335989f3b"
    },
    {
      "type": "bool",
      "concreteTypeId": "b760f44fa5965c2474a3b471467a22c43185152129295af588b022ae50b50903"
    },
    {
      "type": "()",
      "concreteTypeId": "2e38e77b22c314a449e91fafed92a43826ac6aa403ae6a8acb6cf58239fbaf5d"
    }
  ],
  "metadataTypes": [
    {

>       "metadataTypeId": 1,

Reviewer 4: should we make metadataTypeId's string based in case we want to make them hashed later?

```json
{
  "concreteTypes": [
    {
      "type": "([str[5]; 3], bool, b256)",
      "concreteTypeId": "625531542be70834dd127e771101ac1014111718451bfae996d97abe700c66a5",
      "metadataTypeId": 1,
    },
    {
      "type": "[str[5]; 3]",

>       "concreteTypeId": "40c357685306e593eb4c4154377425853a7387ac5a6962d1d9198081a011d64a",

Reviewer: In this example, this concrete type is not referenced anywhere; instead, its referenced metadata type is used [here](https://github.com/FuelLabs/fuel-specs/pull/599/files#diff-29c836429beedee654a72c37d1f883c1e5a0084d468e1446c61da13d7f88783dR407). Does this type have a role? If not, is there place for optimization here so that only the _used_ concrete types are included in the `concreteTypes` array?
